### PR TITLE
[Metricbeat] Testing: Allow use of different directory to test data instead of the expected one

### PR DESCRIPTION
**Pull request by constanca-m**
## What does this PR do?

Offers the possibility to define the read and write folder for the files used for `mbtest.TestDataFiles`. The previous default folder remains the same.

## Why is it important?

At this moment, each metricset that uses `mbtest.TestDataFiles`, needs to have the directory `_meta/testdata`. However, sometimes a metricset can be using the same file as another. See this [change](https://github.com/elastic/beats/pull/34448/files) where multiple metricsets need to have the same files. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Outcome

Example: when running the testing function inside directory `module/kubernetes/state_cronjob`:
```
func TestData(t *testing.T) {
	mbtest.TestDataFiles(t, "kubernetes", name)
}
```

The results should be:
```
=== RUN   TestData
--- PASS: TestData (0.50s)
=== RUN   TestData/ksm.v2.4.2.plain
    testdata.go:216: Testing _meta/testdata/ksm.v2.4.2.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.4.2.plain-expected.json file
    --- PASS: TestData/ksm.v2.4.2.plain (0.12s)
=== RUN   TestData/ksm.v2.5.0.plain
    testdata.go:216: Testing _meta/testdata/ksm.v2.5.0.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.5.0.plain-expected.json file
    --- PASS: TestData/ksm.v2.5.0.plain (0.13s)
=== RUN   TestData/ksm.v2.6.0.plain
    testdata.go:216: Testing _meta/testdata/ksm.v2.6.0.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.6.0.plain-expected.json file
    --- PASS: TestData/ksm.v2.6.0.plain (0.13s)
=== RUN   TestData/ksm.v2.7.0.plain
    testdata.go:216: Testing _meta/testdata/ksm.v2.7.0.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.7.0.plain-expected.json file
    --- PASS: TestData/ksm.v2.7.0.plain (0.12s)
PASS
```

However, if specified a custom prefix in the config.yml file:
```yaml
path: "../_meta/testdata"
```

Running the function:
```
func TestData(t *testing.T) {
	mbtest.TestDataFiles(t, "kubernetes", name)
}
```
Would lead to:
```
=== RUN   TestData
--- PASS: TestData (0.49s)
=== RUN   TestData/ksm.v2.4.2.plain
    testdata.go:216: Testing ../_meta/testdata/ksm.v2.4.2.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.4.2.plain-expected.json file
    --- PASS: TestData/ksm.v2.4.2.plain (0.12s)
=== RUN   TestData/ksm.v2.5.0.plain
    testdata.go:216: Testing ../_meta/testdata/ksm.v2.5.0.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.5.0.plain-expected.json file
    --- PASS: TestData/ksm.v2.5.0.plain (0.14s)
=== RUN   TestData/ksm.v2.6.0.plain
    testdata.go:216: Testing ../_meta/testdata/ksm.v2.6.0.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.6.0.plain-expected.json file
    --- PASS: TestData/ksm.v2.6.0.plain (0.12s)
=== RUN   TestData/ksm.v2.7.0.plain
    testdata.go:216: Testing ../_meta/testdata/ksm.v2.7.0.plain file
    testdata.go:286: Expected _meta/testdata/ksm.v2.7.0.plain-expected.json file
    --- PASS: TestData/ksm.v2.7.0.plain (0.11s)
PASS
```

### CODE REVIEW

**Comment by constanca-m** [comment]
Hey @elastic/elastic-agent-data-plane, can you have a look at this? Thanks.

File: metricbeat/module/openmetrics/collector/collector_test.go [hunk 1]
```diff
@@ -43,23 +43,23 @@ func TestData(t *testing.T) {
...
 		"handler": "query",
 	}
 	tests := []struct {
-		Family *p.MetricFamily
+		Family *prometheus.MetricFamily
 		Event  []OpenMetricEvent
 	}{
 		{
-			Family: &p.MetricFamily{
+			Family: &prometheus.MetricFamily{
```
Comment (tetianakravchenko) [inline metricbeat/module/openmetrics/collector/collector_test.go:58]: are those changes related to this PR?

**Review by tetianakravchenko — COMMENTED** [review]
is it actually required for the purpose of this PR - delete tests from the openmentrics module? I didn't check why do we have 2 folders with tests, but I think that it shouldn't part of this PR, can it be split to 2 PRs: move openmetrics changes to different PR?

**Comment by tetianakravchenko** [comment]
@constanca-m also looking on my comment from your previous PR - https://github.com/elastic/beats/pull/34448#discussion_r1094252085, maybe it would be easier to add a symlink files for the `metricbeat/module/kubernetes/_meta/test/ksm.v2.4.2` instead?

File: metricbeat/module/openmetrics/collector/collector_test.go (hunk 1 above)
Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/collector_test.go:58]: Yes, this needs to be changed in this PR. Otherwise tests will keep failing. I checked and one of the folder was a duplicate, so I deleted it to fix the error.

**Comment by constanca-m** [comment]
> @constanca-m also looking on my comment from your previous PR - [#34448 (comment)](https://github.com/elastic/beats/pull/34448#discussion_r1094252085), maybe it would be easier to add a symlink files for the `metricbeat/module/kubernetes/_meta/test/ksm.v2.4.2` instead?

I am not sure what that implies @tetianakravchenko

File: metricbeat/module/openmetrics/collector/collector_test.go (hunk 1 above)
Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/collector_test.go:58]: @constanca-m I don't think it is a duplicate, module configuration is different - in `testdata` https://github.com/elastic/beats/blob/3268f335f0d07c306d9c38ab1f96dc4498a8a0c8/metricbeat/module/openmetrics/collector/_meta/testdata/config.yml#L6-L8 
compare to `samelabeltestdata` - https://github.com/elastic/beats/blob/3268f335f0d07c306d9c38ab1f96dc4498a8a0c8/metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/config.yml#L6-L8

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/collector_test.go:58]: `testdata` folder is not being used anywhere

Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/collector_test.go:58]: > testdata folder is not being used anywhere

what do you mean by this? tests were running only for `samelabeltestdata`, or ?

this extra tests folder with different module configuration was added in this PR https://github.com/elastic/beats/pull/27269. @ChrsMark maybe you remember why/if here we need to keep both tests with different configurations?

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/collector_test.go:58]: > tests were running only for `samelabeltestdata`

Yes

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -58,7 +58,7 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 
 
 *Filebeat*
-- [Auditbeat System Package] Added support for Apple Silicon chips. {pull}34433[34433] 
+- [Auditbeat System Package] Added support for Apple Silicon chips. {pull}34433[34433]
```
Comment (ChrsMark) [inline CHANGELOG.next.asciidoc:61]: nit: is this change needed? If not please revert.

File: metricbeat/module/openmetrics/collector/collector_test.go (hunk 1 above)
Comment (ChrsMark) [reply metricbeat/module/openmetrics/collector/collector_test.go:58]: From what I can see, these settings have to do with how the data are grouped at https://github.com/elastic/beats/blob/9754d2b04f661e797cae113116785509d7d5ec89/metricbeat/module/openmetrics/collector/collector.go#L160-L164. So this is why there are 2 different testdata folders.

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/config.yml [hunk 3]
```diff
@@ -1,8 +0,0 @@
-type: http
```
Comment (ChrsMark) [inline metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/config.yml:1]: I don't think this directory should be removed. It's an actual test case that had been working so far, right?
Unless there is strong reason for it, we should not remove it.

File: metricbeat/mb/testing/testdata.go [hunk 4]
```diff
@@ -150,14 +151,29 @@ func TestDataFiles(t *testing.T, module, metricSet string) {
...
 // TestDataFilesWithConfig run tests for a testdata config
-func TestDataFilesWithConfig(t *testing.T, module, metricSet string, config DataConfig) {
+func TestDataFilesWithConfig(t *testing.T, module, metricSet string, config DataConfig, testPrefix ...string) {
 	t.Helper()
-	ff, err := filepath.Glob(filepath.Join(config.Path, "*."+config.Suffix))
+
+	// the location of the read files
+	location := filepath.Join(config.Path, "*."+config.Suffix)
+
+	// if this function was called from data_test.go then the testPrefix should be 1
```
Comment (ChrsMark) [inline metricbeat/mb/testing/testdata.go:160]: `testPrefix` is a string that holds the directory prefix, right? How it can 1?

File: metricbeat/mb/testing/testdata.go [hunk 5]
```diff
@@ -193,6 +209,13 @@ func TestMetricsetFieldsDocumented(t *testing.T, metricSet mb.MetricSet, events
 }
 
 func runTest(t *testing.T, file string, module, metricSetName string, config DataConfig) {
+	filename := filepath.Base(file)
```
Comment (ChrsMark) [inline metricbeat/mb/testing/testdata.go:212]: Why is it needed now? What was different in the past? Was the `data.json` included in the testing somehow?

Could you add an explanatory comment of what this code block is trying to achieve?

File: metricbeat/mb/testing/testdata.go [hunk 6]
```diff
@@ -242,22 +265,25 @@ func runTest(t *testing.T, file string, module, metricSetName string, config Dat
 			err, module, metricSetName)
 	}
 
+	expectedFile := filepath.Join(config.WritePath, filename+expectedExtension)
```
Comment (ChrsMark) [inline metricbeat/mb/testing/testdata.go:268]: Can you just do the concatenation only if the `WritePath` is not empty? In this, the code will imply that `WritePath` is sth optional.

File: metricbeat/mb/testing/testdata.go [hunk 7]
```diff
@@ -150,14 +151,29 @@ func TestDataFiles(t *testing.T, module, metricSet string) {
...
+	location := filepath.Join(config.Path, "*."+config.Suffix)
+
+	// if this function was called from data_test.go then the testPrefix should be 1
+	if len(testPrefix) != 0 {
+		// the prefix for read and write files should be ../../../module/moduleName/metricsetName
+		prefix := filepath.Join(testPrefix[0], module, metricSet)
+		location = filepath.Join(prefix, location)
+
+		// the prefix needs to be appended to the path of the expected files and the original files
+		config.WritePath = filepath.Join(prefix, config.WritePath)
```
Comment (ChrsMark) [inline metricbeat/mb/testing/testdata.go:167]: That really worries me that we override the configuration. Can we avoid this please?

**Review by ChrsMark — CHANGES_REQUESTED** [review]
The addition in the test helper looks good in general if it is to reduce code/file duplication. I didn't check it in detail but if the testing suite doesn't break it should be fine. I left some nits to be considered to improve the changes a bit.

The removed files from the `openmetrics` module should not be removed, those are actual unit tests.

These changes should also be depicted at https://github.com/elastic/beats/blob/main/metricbeat/mb/testing/data/README.md.

File: CHANGELOG.next.asciidoc (hunk 2 above)
Comment (constanca-m) [reply CHANGELOG.next.asciidoc:61]: I think they were automatic. But what is the problem in removing one space?

File: metricbeat/mb/testing/testdata.go (hunk 4 above)
Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:160]: Not sure I understand. If this function was called from the test to check that testing is ok (outside the modules), then there is a prefix (see [this](https://github.com/elastic/beats/blob/9d7e4c9ac133e9b578a58848d4f8177b0676f0d4/metricbeat/mb/testing/data/data_test.go#L52)).

File: metricbeat/mb/testing/testdata.go (hunk 5 above)
Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:212]: > Was the data.json included in the testing somehow?

Yes.

We also were not using the custom write path, although it should be possible before. I'm guessing it was a distraction.

File: metricbeat/mb/testing/testdata.go (hunk 6 above)
Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:268]: I guess, but by default `writepath` is never empty.

File: metricbeat/mb/testing/testdata.go (hunk 4 above)
Comment (MichaelKatsoulis) [reply metricbeat/mb/testing/testdata.go:160]: In the comments it says `if this function was called from data_test.go then the testPrefix should be 1` . How can it be 1?

Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:160]: I am still confused, sorry. The line that I linked is this:
`mbtest.TestDataFilesWithConfig(t, moduleName, metricSetName, config, getModulesPath())`
So you can see that there is an extra parameter being passed. That is why the []string has length 1.

Comment (ChrsMark) [inline metricbeat/mb/testing/testdata.go:160]:
```suggestion
	// if this function was called from data_test.go then the length of testPrefix should greater than 0.
```
Maybe you mean sth like this?

Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:160]: It is also correct, but it will always be one:

```
func getModulesPath() string {
	return "../../../module"
}
```

Why not be specific?

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/config.yml (hunk 3 above)
Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/config.yml:1]: I will fix that. I still think one of the 2 is not being used, but maybe that is a change for a different PR.

File: CHANGELOG.next.asciidoc (hunk 2 above)
Comment (ChrsMark) [reply CHANGELOG.next.asciidoc]: In principle a PR should not introduce changes that are not intentional, even if it is just one space or an empty line.
For this one I think that the editor is correct and removes the space that is not needed, so you can leave it as if you prefer since it is an improvement :).

File: metricbeat/mb/testing/testdata.go (hunk 4 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:160]: I don't think I get it 😞 Could you share what is the actual value of the `testPrefix`, or an example of what it would be?

File: metricbeat/mb/testing/testdata.go (hunk 7 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:167]: Maybe it would be nicer if those are initialised inside the "constructor" that initialised the default config -> https://github.com/elastic/beats/blob/89bcc33a9a90cbacceaa61fceec0e8d23073574c/metricbeat/mb/testing/testdata.go#L112

File: metricbeat/mb/testing/testdata.go (hunk 5 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:212]: > Yes.

How? I'm not about this change and how it affects the end result, could you elaborate more on this on why we exclude this file and what was the behavior before?

Then it would be nice to have this as a comment in the code please so as to know what these lines intend to do.

File: metricbeat/mb/testing/testdata.go (hunk 4 above)
Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:160]: Of course. In the `data_test.go` file we have two functions: `TestAll` and `getModulesPath`. For this specific parameter they are like this:
```
func TestAll(t *testing.T) {
   ...
      mbtest.TestDataFilesWithConfig(t, moduleName, metricSetName, config, getModulesPath())
   ...
}

func getModulesPath() string {
	return "../../../module"
}
```

And we can see that the `TestDataFilesWithConfig` function expects this parameters:
```
func TestDataFilesWithConfig(t *testing.T, module, metricSet string, config DataConfig, testPrefix ...string)
```

From this, we know that when this function is called from `data_test.go`:
```
mbtest.TestDataFilesWithConfig(t, moduleName, metricSetName, config, "../../../module")
```

The string has length 1: ["../../../module"].

If we know look into how this function is called from the modules (example from `collector_test.go`):
```
func TestSameLabels(t *testing.T) {
	dataConfig := mbtest.ReadDataConfig(t, "_meta/samelabeltestdata/config.yml")
	mbtest.TestDataFilesWithConfig(t, "openmetrics", "collector", dataConfig)
}
```
We can see that we only have 4 parameters. So our array will have length 0 (this happens with all modules).

File: metricbeat/mb/testing/testdata.go (hunk 6 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:268]: Got it, thanks!

File: metricbeat/mb/testing/testdata.go (hunk 7 above)
Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:167]: We don't have access to the prefix when that function is called, that is why we need to change it at that moment.

File: metricbeat/mb/testing/testdata.go (hunk 5 above)
Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:212]: Yes, I wrote this comment to explain it:
```
If the expected suffix is '.json' we need to exclude the data.json file, since by the end of this function we may create a data.json file if there is a docs file with the config suffix:
			if strings.HasSuffix(file, "docs."+config.Suffix) {
				writeDataJSON(t, data[0], filepath.Join(config.WritePath, "data.json"))
			}
Since the expected file name is obtained through filename + expectedExtension, we could end up testing files like:
			Metrics file: data.json
			Expected metrics file: data.json-expected.json
If the config extension is '.json'.
This is not possible, since running go -test json does not produce an expected file for data.json files. This is why we need to exclude this file from the tests.
```
This is a mistake that wasn't detected before, because `writePath` was never being used.

File: metricbeat/mb/testing/testdata.go (hunk 7 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:167]: Got it, thanks for the extra info!

File: metricbeat/mb/testing/testdata.go (hunk 4 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:160]: I see why it was confusing, you pass the variable as an array with `testPrefix ...string`, but why not to pass it as `string` like `testPrefix string`? It's less confusing it seems :)

Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:160]: It is not possible to pass a string as optional, that is why I put it as an array.

File: metricbeat/mb/testing/testdata.go (hunk 5 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:212]: Ok since it fixes an existent issue.

File: metricbeat/mb/testing/testdata.go (hunk 4 above)
Comment (ChrsMark) [reply metricbeat/mb/testing/testdata.go:160]: Hmm, then maybe you cannot make it optional and pass a `""` an empty string in the calls where it is actually empty? This is less confusing when following the code flow I think.

Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go:160]: That makes it necessary to change every call in each module (it is used). Is this the best way?

Comment (constanca-m) [reply metricbeat/mb/testing/testdata.go]: Ok, I made it mandatory, I am trusting your criteria :)

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (tetianakravchenko) [inline metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: what is the reason of generation `data.json` now? for both `samelabeltestdata` and `testdata`?

**Comment by tetianakravchenko** [comment]
could you please add information about settings `path` and `writepath` to https://github.com/elastic/beats/blob/main/metricbeat/mb/testing/data/README.md#available-settings-in-configyml, and what are the default values

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json (hunk 8 above)
Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: I don't know, that was already in the code. It was added because I ran the tests locally.

**Comment by constanca-m** [comment]
> could you please add information about settings `path` and `writepath` to https://github.com/elastic/beats/blob/main/metricbeat/mb/testing/data/README.md#available-settings-in-configyml, and what are the default values

Yes, good idea @tetianakravchenko

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json (hunk 8 above)
Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: > that was already in the code.

what does it mean? those files were not in `samelabeltestdata` and `testdata` folders before your changes. And as I see in other modules this file is not present (example - https://github.com/elastic/beats/tree/main/metricbeat/module/prometheus/collector/_meta/testdata, similar files structure with the `openmetrics` module)

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: Sorry, should have been more clear. There is [these lines](https://github.com/elastic/beats/blob/79f48040f746dd5e48f951a3a80e197776a9543e/metricbeat/mb/testing/testdata.go#L348-L350) to create that file. Since we ran the tests locally, the files were created.

Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: if running tests locally from main branch (as described in this [doc](https://github.com/elastic/beats/blob/79f48040f746dd5e48f951a3a80e197776a9543e/metricbeat/mb/testing/data/README.md)) `data.json` are not created on the metricsets level, please double check why it gets created after your changes

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: That is happening because `config.WritePath` was not being correctly used before: check [this line](https://github.com/elastic/beats/blob/9f15870697559dfbf48601ca59b56588102b0c87/metricbeat/mb/testing/testdata.go#L115), that defaults to empty string. And from [here](https://github.com/elastic/beats/blob/9f15870697559dfbf48601ca59b56588102b0c87/metricbeat/mb/testing/testdata.go#L309-L311), the file `data.json` is saved under `./meta/data.json`. This is wrong, because if we have more that one folder with files `docs.(config.Sufix)`, we are overwriting them instead of creating the correct ones.

Comment (gizas) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: So I removed data.json under testdata, I changed parent folder to be module/openmetrics/collector and run test.

And the data.json was created under testdata folder (of course another data.json was. created under samelabetestdata.) Thanks for clarification @constanca-m . Looks better to me now

I wonder if we need to document this

Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: it should impact then all other modules/metricsets as well, right?

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: Maybe, but since there were no problems in the documentation until now, maybe the other teams responsible for the other modules aren't minding the problem. And if I do that change in this PR, then I will have to check them all and change the references in the docs.

Comment (gizas) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: Yes if they expect the data.json to other folder. Probably we only check the outcome of the tests only and not where the data.json folder is saved. But this is my assumption

**Comment by gizas** [comment]
> ```yaml
> "../_meta/testdata"
> ```

Why dont we put specific examples in the README. Always an example talks better. An example with or without writepath can be nice

**Comment by gizas** [comment]
For my understanding, our we going to change the _meta/testdata of all our datastreams to a common folder and update the confi.yml files, as part of another PR?

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json (hunk 8 above)
Comment (ChrsMark) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: Keep in mind that `data.json` is automatically attached in the docs page of each metricset. For example: https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-redis-info.html. Not sure if this PR affects this, but worth verifying.

**Comment by constanca-m** [comment]
> For my understanding, our we going to change the _meta/testdata of all our datastreams to a common folder and update the confi.yml files, as part of another PR?

No, the folders don't need update nor the config files. The only thing this changes is that the `data.json` file is going to be written in the expected folder, instead of the default one. I wasn't thinking in changing the `data.json` files since they are working where they are. @gizas

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json (hunk 8 above)
Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: It doesn't, because the files are still there. Refactoring the test doesn't delete them.

Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: So `metricbeat/module/openmetrics/collector/_meta/data.json` will still be use for documentation (doc page of [openmetrics module](https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-openmetrics-collector.html#_fields_190)) ?
but `samelabeltestdata/data.json` or `testdata/data.json` will not be used anywhere in the public documentation?

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: Yes, that is correct @tetianakravchenko

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: The data.json file is probably being used in a wrong way, since there are two folders writing on it.

Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: I am not sure if we should store it in repo then (the one under `_meta/testdata/data.json`) if it is not used anywhere 🤔 

Is `metricbeat/module/moduleName/metricsetName/_meta/data.json` updates automatically after your changes? since this location will be used for the public documentation, this writePath for data.json should be supported

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: > I am not sure if we need to store it in repo then (the one under _meta/testdata/data.json) if it is not used anywhere thinking or am I missing smth here?

No, you're right. But if someone decided to write `data.json` files, it is because they are important, I assume?

> Is metricbeat/module/moduleName/metricsetName/_meta/data.json updates automatically after your changes? since this location will be used for the public documentation, this writePath for data.json should be supported

Yes, but keep in mind this was the behavior that was expected before. But since there were no other modules using two folders for writing this file, this behavior went unnoticed.

Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: > > Is metricbeat/module/moduleName/metricsetName/_meta/data.json updates automatically after your changes? since this location will be used for the public documentation, this writePath for data.json should be supported
>
> Yes, but keep in mind this was the behavior that was expected before. But since there were no other modules using two folders for writing this file, this behavior went unnoticed.

sorry, I didn't understand. so update of `metricbeat/module/moduleName/metricsetName/_meta/data.json` will not be supported after merging this PR? and for some modules it implies that future changes of fields (or adding new module) will not be reflected in `metricbeat/module/moduleName/metricsetName/_meta/data.json` -> hence no changes in documentation?

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: > sorry, I didn't understand. so update of metricbeat/module/moduleName/metricsetName/_meta/data.json will not be supported after merging this PR? and for some modules it implies that future changes of fields (or adding new module) will not be reflected in metricbeat/module/moduleName/metricsetName/_meta/data.json -> hence no changes in documentation?

Yes, that file will no longer be updated. But that is not exactly a problem, unless the fields change like you mentioned.

**Comment by gizas** [comment]
I was referring to a common folder eg. for all state_datasets. Maybe this way we can avoid repetition. And again yes only to another PR.

**Comment by constanca-m** [comment]
> I was referring to a common folder eg. for all state_datasets. Maybe this way we can avoid repetition. And again yes only to another PR. But ok not something important

I think the `data.json` is different for all `state_*` because they have different fields.

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json (hunk 8 above)
Comment (tetianakravchenko) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: it will be a breaking change for our documentation, it should be changed

@elastic/elastic-agent-data-plane could you please review?

**Review by tetianakravchenko — CHANGES_REQUESTED** [review]

File: metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json (hunk 8 above)
Comment (cmacknz) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: If I am interpreting this correctly, we are proposing to stop updating the data.json file that is used to render example events published in our official documentation? For example https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-redis-info.html that was already linked?

If this understanding is correct then,

1. We should keep the data.json field up to date. Users are likely relying on these pages to know what fields exist in the events for querying and processing. Removing this isn't helping our users, even if it may be convenient for us.
2. If we have discovered that data.json is not being maintained or generated consistently, then that seems like a bug causing the documentation to be out date and should be fixed.

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: > If I am interpreting this correctly, we are proposing to stop updating the data.json file that is used to render example events published in our official documentation? For example https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-redis-info.html that was already linked?

Yes, you're correct. But the way we are referencing these files is wrong. And the one that you linked is not affected by this, because there are no files referenced there generated by this function. But going back to the ones that are actually affected... We can see a good example in this changed module `openmetrics`. We have two new `data.json` files: one for `samelabeltestdata` (created by running [this test](https://github.com/constanca-m/beats/blob/79f48040f746dd5e48f951a3a80e197776a9543e/metricbeat/module/openmetrics/collector/collector_test.go#L41-L43)) and another for `testdata` (created by running [this test](https://github.com/constanca-m/beats/blob/79f48040f746dd5e48f951a3a80e197776a9543e/metricbeat/module/openmetrics/collector/collector_test.go#L45-L48)). Previously, running these two tests would result in just one `data.json` file (running one function after the other, is overwriting what the other other did), misplaced in [this directory](https://github.com/constanca-m/beats/tree/refactor-test-data-mb/metricbeat/module/openmetrics/collector/_meta).

So if:
1. We want to always create just `data.json` file and never have the possibility to create more than one for different configs, this PR is breaking that.
2. If we want to fix it, this PR is needed for that, but that requires changing the path for the `data.json` files in some of the modules. If this is not done, the referenced `data.json` files in the docs will stop being updated.

I would say that as it is now, this is a bug and we should fix it.

Comment (cmacknz) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json:1]: I went and looked at how the data.json files are actually pulled into our documentation. Picking the Redis info one that is already referenced (arbitrarily, I know it isn't affected) the data.json file is included in the docs with this asciidoc syntax:

https://github.com/elastic/beats/blob/0d90255102319e88afbaba7aac33c3f8cf05a88a/metricbeat/docs/modules/redis/info.asciidoc#L16-L27

You can see that we are explicitly linking to the data.json file to generate an example event. 

I think the only two requirements here are: 
1. We can generate sample events for the metricsets in each module. 
2. We do not remove or break the example events that are already in our documentation.

We can change the name of the .json file and the number of .json files generated as long as we update the links in the documentation to account for that.

> We want to always create just one data.json file and never have the possibility to create more than one for different configs, this PR is breaking that

I don't think only having one data.json file is a requirement. We can link to multiple different files as needed.

> If we want to fix it, this PR is needed for that, but that requires changing the path for the data.json files in some of the modules. If this is not done, the referenced data.json files in the docs will stop being updated

We can change the path to the generated files as needed, we just need to also update the relative paths referenced in the .asciidoc files that are used to generate the documentation as well.

Comment (constanca-m) [reply metricbeat/module/openmetrics/collector/_meta/samelabeltestdata/data.json]: > You can see that we are explicitly linking to the data.json file to generate an example event.

Yes, I noticed. To avoid changing the path then (I think this would be very simple, we would just need to change [this part of the template](https://github.com/elastic/beats/blob/f51d8e6cd68a1f551910417023f7cd615dec19ca/metricbeat/scripts/mage/template/metricsetDoc.tmpl#L26-L37)), I set a default one specifically for the `data.json` file. By default this path is always the same, so if there are two tests overwritting this file (like it is happening with the module openmetrics), at least there is a chance now to change the path for one of the files if it is needed.

So this way we keep updating the old `data.json` files and give an option if we want to create more than one.

File: metricbeat/mb/testing/data/README.md [hunk 9]
```diff
@@ -26,7 +26,8 @@ An alternative is to just run from metricbeat `mage mockedTests` to achieve the
 ### Available settings in `config.yml`
 
 - `path`: (string) Path to reach the directory containing the files to read from. Default is "_meta/testdata".
-- `wirtepath`: (string) Path to the directory where the expected files are going to be written. Default is "_meta/testdata".
+- `writepath`: (string) Path to the directory where the expected files are going to be written. Default is "_meta/testdata".
+- `datapath`: (string) Path to the directory where the data.json file is going to be written. Default is "_meta".
```
Comment (tetianakravchenko) [inline metricbeat/mb/testing/data/README.md:30]: could you please somewhere add a comment why we have `data.json` file and that is used/might be used for our public documentation (or link to the place where it is mentioned, but I couldn't find it)

Comment (constanca-m) [reply metricbeat/mb/testing/data/README.md:30]: I can try, but I am not sure I understand the reasoning for that file since it was already there

Comment (tetianakravchenko) [reply metricbeat/mb/testing/data/README.md:30]: I was thinking about adding smth like `data.json file is used to render example events published in our official documentation`
at least for me it wasn't clear at the moment of asking why we generated those files in this comment https://github.com/elastic/beats/pull/34467#discussion_r1101788362

Comment (constanca-m) [reply metricbeat/mb/testing/data/README.md]: Ok, done that! I was thinking something more complex, but as you wrote it, it makes perfect sense. I don't want to link to the line code because I would have to use one fixed commit and that might end up not making sense if we are on the branch main and are sent to a possible very old one.

**Review by tetianakravchenko — APPROVED** [review]
LGTM!

**Review by ChrsMark — APPROVED** [review]
LGTM!
Make sure you have a green CI before merging!
